### PR TITLE
Add 'index_NS', make interface more systematic

### PR DESCRIPTION
**Pull request by kosmikus**
Trying to implement the results of the discussion in #22.

### CODE REVIEW

File: src/Generics/SOP/Classes.hs [hunk 1]
```diff
@@ -336,18 +376,90 @@ class HAp h => HSequence (h :: (k -> *) -> (l -> *)) where
...
+  -- >>> hindex (S (S (Z (I False))))
+  -- 2
+  -- >>> hindex (Z (K ()))
+  -- 0
+  -- >>> hindex (SOP (S (Z (I True :* I 'x' :* Nil))))
+  -- 1
+  --
+  -- @since 0.2.4.0
+  --
+  hindex :: h f xs -> Int
```
Comment (phadej) [inline src/Generics/SOP/Classes.hs]: Came to mind that `NS (K ()) xs` and `Fin (length xs) ≈ Int` are isomorphic. I'm not sure I needed (or whether it's useful) to have

```haskell
hunindex :: SList xs => Int -> (forall a. f a) -> h f xs
hunindex n x = hapInjs (hpure x) !! n

hunindexc :: (SList xs, All c xs) => Proxy c => Int -> (forall a. c a => f a) -> h f xs
hunindexc n c x = hapInjs (hcpure c x) !! n
```

directly available?

File: src/Generics/SOP/Classes.hs [hunk 2]
```diff
@@ -336,18 +376,90 @@ class HAp h => HSequence (h :: (k -> *) -> (l -> *)) where
...
+-- | A class for applying all injections corresponding to a sum-like
+-- structure to a table containing suitable arguments.
+--
+class (UnProd (Prod h) ~ h) => HApInjs (h :: (k -> *) -> (l -> *)) where
+
+  -- | For a given table (product-like structure), produce a list where
+  -- each element corresponds to the application of an injection function
+  -- into the corresponding sum-like structure.
+  --
+  -- /Instances:/
```
Comment (phadej) [inline src/Generics/SOP/Classes.hs]: FYI: haddock supports

```
=== Instances
```

headers.

File: src/Generics/SOP/Classes.hs (hunk 1 above)
Comment (kosmikus) [inline src/Generics/SOP/Classes.hs]: Interesting. I'll think about this, but as you said, this shouldn't block the current PR.

File: src/Generics/SOP/Classes.hs (hunk 2 above)
Comment (kosmikus) [inline src/Generics/SOP/Classes.hs]: Is there any special treatment of this? Or is it just a different form of markup to use?

Comment (phadej) [inline src/Generics/SOP/Classes.hs]: Different form of markup, but it has nice feature: https://github.com/haskell/haddock/issues/335

But again, just FYI. Those "subsections" aren't annoyingly long so they'd benefit from collapsing. And it's up to your preference, whether you want "header" styling :)

Comment (kosmikus) [inline src/Generics/SOP/Classes.hs]: Yes, ok. Having examples and instances collapsible is perhaps an improvement. I'll see if I can do this in another sweep for the whole package.

**Pull request merged**